unit-tests/src/org/commcare/android/tests/formnav/EndOfFormTest.java

```diff
@@ -0,0 +1,143 @@
+
+        TestUtils.initializeStaticTestStorage();
+        TestAppInstaller.setupPrototypeFactory();
+
+        TestAppInstaller appTestInstaller =
+                new TestAppInstaller("jr://resource/commcare-apps/form_nav_tests/profile.ccpr",
+                        "test", "123");
+        appTestInstaller.installAppAndLogin();
+        ShadowEnvironment.setExternalStorageState(Environment.MEDIA_MOUNTED);
+    }
+
+    /**
+     * Test filling out and saving a form ending in a hidden repeat group. This
+     * type of form exercises uncommon end-of-form code paths
+     */
+    @Test
+    public void testHiddenRepeatAtEndOfForm() {
+        CommCareHomeActivity homeActivity = buildHomeActivityForFormEntryLaunch();
+
+        ShadowActivity shadowActivity = Shadows.shadowOf(homeActivity);
+        Intent formEntryIntent = shadowActivity.getNextStartedActivity();
+
+        // make sure the form entry activity should be launched
+        String intentActivityName = formEntryIntent.getComponent().getClassName();
+        assertTrue(intentActivityName.equals(FormEntryActivity.class.getName()));
+
+        ShadowActivity shadowFormEntryActivity = navigateFormEntry(formEntryIntent);
+
+        // trigger CommCareHomeActivity.onActivityResult for the completion of
+        // FormEntryActivity
+        shadowActivity.receiveResult(formEntryIntent,
+                shadowFormEntryActivity.getResultCode(),
+                shadowFormEntryActivity.getResultIntent());
+        assertStoredFroms();
+    }
+
+    private CommCareHomeActivity buildHomeActivityForFormEntryLaunch() {
+        AndroidSessionWrapper sessionWrapper =
+                CommCareApplication._().getCurrentSessionWrapper();
+        CommCareSession session = sessionWrapper.getSession();
+        session.setCommand("m0-f0");
+
+        CommCareHomeActivity homeActivity =
+                Robolectric.buildActivity(CommCareHomeActivity.class).create().get();
+        // make sure we don't actually submit forms by using a fake form submitter
+        homeActivity.setFormAndDataSyncer(new FormAndDataSyncerFake(homeActivity));
+        SessionNavigator sessionNavigator = homeActivity.getSessionNavigator();
+        sessionNavigator.startNextSessionStep();
+        return homeActivity;
+    }
+
+    private ShadowActivity navigateFormEntry(Intent formEntryIntent) {
+        // launch form entry
+        FormEntryActivity formEntryActivity =
+                Robolectric.buildActivity(FormEntryActivity.class).withIntent(formEntryIntent)
+                        .create().start().resume().get();
+
+        ImageButton nextButton = (ImageButton)formEntryActivity.findViewById(R.id.nav_btn_next);
+        // demonstrate form progress bug where it thinks the form is done until
+        // you enter an answer for the question, which effects the hidden
+        // repeat coming up next
+        assertTrue(nextButton.getTag().equals(FormEntryActivity.NAV_STATE_DONE));
```

Comment:
It is a hard bug to fix. I don't really know why I'm highlighting it in this tests; I guess as a personal note because I intend to eventually look into it and this is a good example of how to reproduce it.